### PR TITLE
Support for dynamodbstreams

### PR DESCRIPTION
**Pull request by dmarjenburgh**
Basic low-level support for dynamodb-streams.

### CODE REVIEW

**Comment by dmarjenburgh** [comment]
I just thought I should probably add some documentation to the README.

Also, let me know if you want me to squash the commits into one.

**Comment by ptaoussanis** [comment]
> Also, let me know if you want me to squash the commits into one.

Thanks, ideally wouldn't mind squashing the Cleanup commit into the relevant source commits. Aside from that, looks good as is.

Also about to make one or two small inline comments...

File: src/taoensso/faraday.clj [hunk 1]
```diff
@@ -369,15 +411,18 @@
...
-     :status        (utils/un-enum (.getTableStatus d))
+    {:name                 (keyword (.getTableName d))
+     :creation-date        (.getCreationDateTime d)
+     :item-count           (.getItemCount d)
+     :size                 (.getTableSizeBytes d)
+     :throughput           (as-map (.getProvisionedThroughput d))
+     :indexes              (as-map (.getLocalSecondaryIndexes d)) ; DEPRECATED
+     :lsindexes            (as-map (.getLocalSecondaryIndexes d))
+     :gsindexes            (as-map (.getGlobalSecondaryIndexes d))
+     :stream-specification (as-map (.getStreamSpecification d))
```
Comment (ptaoussanis) [inline src/taoensso/faraday.clj]: Perhaps `:stream-spec` would be more convenient + sufficiently unambiguous?

To clarify: suggesting a change of this and all other API uses of `:stream-specification`.

File: test/taoensso/faraday/tests/requests.clj [hunk 2]
```diff
@@ -6,92 +6,100 @@
...
-    ConditionalCheckFailedException
-    DeleteGlobalSecondaryIndexAction
-    InternalServerErrorException
-    ItemCollectionSizeLimitExceededException
-    LimitExceededException
-    ProvisionedThroughputExceededException
-    ResourceInUseException
-    ResourceNotFoundException
-    UpdateGlobalSecondaryIndexAction]))
+    [com.amazonaws.services.dynamodbv2.model
```
Comment (ptaoussanis) [inline test/taoensso/faraday/tests/requests.clj]: May I request that the indenting not be changed here? Just creating some unnecessary diff noise. Thanks!

File: test/taoensso/faraday/tests/requests.clj [hunk 3]
```diff
@@ -219,6 +233,13 @@
     (expect "global-secondary" (.getIndexName action))
     ))
 
+(let [req ^UpdateTableRequest (update-table-request
```
Comment (ptaoussanis) [inline test/taoensso/faraday/tests/requests.clj]: Please note that type hints should be attached to the symbol; so in this case

```clojure
[^UpdateTableRequest req (update-table-request ...)]
```

File: test/taoensso/faraday/tests/requests.clj [hunk 4]
```diff
@@ -450,3 +471,37 @@
   (expect "age-index" (.getIndexName req))
   (expect (.getConsistentRead req))
   )
+
+(let [req ^ListStreamsRequest (list-stream-request
```
Comment (ptaoussanis) [inline test/taoensso/faraday/tests/requests.clj]: Same comment re: type hints. Will skip commenting on further examples if they come up.

File: test/taoensso/faraday/tests/requests.clj [hunk 5]
```diff
@@ -450,3 +471,37 @@
...
+  (expect ShardIteratorType/AFTER_SEQUENCE_NUMBER (ShardIteratorType/fromValue (.getShardIteratorType req)))
+  (expect "000001" (.getSequenceNumber req)))
+
+(let [req ^GetRecordsRequest (get-records-request
+                               "arn:aws:dynamodb:us-west-2:111122223333:table/etcetc"
+                               {:limit 50})]
+  (expect "arn:aws:dynamodb:us-west-2:111122223333:table/etcetc" (.getShardIterator req))
+  (expect 50 (.getLimit req)))
+
+(test/run-tests '[taoensso.faraday.tests.requests])
```
Comment (dmarjenburgh) [inline test/taoensso/faraday/tests/requests.clj]: I forgot about this line. Will remove it

File: test/taoensso/faraday/tests/requests.clj (hunk 3 above)
Comment (ptaoussanis) [inline test/taoensso/faraday/tests/requests.clj]: Type hints only work when attached to the symbol, otherwise they're ignored. If other tests have the incorrect form then the type hints are either unnecessary or need to be fixed; will check in a moment. Thanks for the report!

File: src/taoensso/faraday.clj (hunk 1 above)
Comment (ptaoussanis) [inline src/taoensso/faraday.clj]: The AWS SDKs can be incredibly verbose. My general preference is to use the shortest name that's unambiguous and intelligible.

I'll note that the rest of Faraday uses this convention, so the suggested change would also help to maintain consistency with the rest of the API.

File: test/taoensso/faraday/tests/requests.clj (hunk 3 above)
Comment (dmarjenburgh) [inline test/taoensso/faraday/tests/requests.clj]: Attaching the type-hint to either the symbol or the form makes the reflection-warnings go away.

Comment (ptaoussanis) [inline test/taoensso/faraday/tests/requests.clj]: Interesting, I stand corrected. This may be a compiler change that happened sometime without me noticing, or I may just be misremembering; was under the incorrect impression that type metadata wasn't in general inferred from a form to a local binding. Thanks for pointing this out!

File: src/taoensso/faraday.clj (hunk 1 above)
Comment (ricardojmendez) [inline src/taoensso/faraday.clj]: My initial thinking was along the lines suggested by @dmarjenburgh, but I kept addition names short for consistency. I keep thinking we can do a mass rename eventually, for a breaking 2.0, but would agree that `:stream-spec` is closer to what DynamoDB has been using.

**Comment by ricardojmendez** [comment]
I can't actually comment on implementation specifics, since I haven't used DynamoDB Streams, but overall the PR looks good.

**Comment by ptaoussanis** [comment]
Hi Daniel, thanks again for the PR (and for your patience on this). Merging now.
